### PR TITLE
Fix incorrect rounding for instant_micros in ORC

### PR DESCRIPTION
**Pull request by martint**
## Description

The picosecond of millisecond fraction can round up to a whole
millisecond. This needs to be accounted for by adjusting the
millisecondEpoch portion.

Not sure how to test this, as I don't think we have infrastructure for generating ORC files with the instant_micros values that will trigger this issue.

> Is this change a fix, improvement, new feature, refactoring, or other?

Bug fix

> Is this a change to the core query engine, a connector, client library, or the SPI interfaces? (be specific)

Connector: Iceberg and Hive

> How would you describe this change to a non-technical end user or system administrator?

## Documentation

(x) No documentation is needed.

## Release notes
```markdown
# Iceberg, Hive
* Fix potential query failure when reading `timestamp(6) with time zone` values. ({issue}`12804`)
```

### CODE REVIEW

**Review by dain — APPROVED** [review]

File: lib/trino-orc/src/main/java/io/trino/orc/reader/TimestampColumnReader.java [hunk 1]
```diff
@@ -616,6 +617,11 @@ private void readInstantMicros(int i, long[] millisValues, int[] picosFractionVa
 
             // round nanos to micros and convert to picos
             picosFraction = toIntExact(roundDiv(nanos, NANOSECONDS_PER_MICROSECOND)) * PICOSECONDS_PER_MICROSECOND;
+
+            if (picosFraction == PICOSECONDS_PER_MILLISECOND) {
+                picosFraction = 0;
+                millis++;
+            }
```
Comment (findepi) [inline lib/trino-orc/src/main/java/io/trino/orc/reader/TimestampColumnReader.java:624]: is this the only place like this in ORC time/timestamp readers?

**Review by findepi — APPROVED** [review]

**Comment by findepi** [comment]
> Not sure how to test this, as I don't think we have infrastructure for generating ORC files with the instant_micros values that will trigger this issue.

Can it happen if we write with Hive with nanosecond precision and read back with microsecond precision?

File: lib/trino-orc/src/main/java/io/trino/orc/reader/TimestampColumnReader.java (hunk 1 above)
Comment (ademille) [reply lib/trino-orc/src/main/java/io/trino/orc/reader/TimestampColumnReader.java:624]: Would readInstantNanos have the same problem? I guess it doesn't include the rounding code, so it's probably ok.

**Comment by martint** [comment]
@findepi, added a test. Please take a look.

File: lib/trino-orc/src/main/java/io/trino/orc/reader/TimestampColumnReader.java (hunk 1 above)
Comment (martint) [reply lib/trino-orc/src/main/java/io/trino/orc/reader/TimestampColumnReader.java:624]: That method is safe, since it doesn't need to round.

**Review by findepi — APPROVED** [review]

**Review by dain — APPROVED** [review]

**Pull request merged**